### PR TITLE
Added DataFileValue to SetEqualityIterator checks in conditional mutations

### LINKED ISSUE #3522: Explore comparing file values in requireSame
In #3513 a new requireSame conditional mutation check was added to Ample.  This check allows easily comparing if something read from a tablet is the same before making an update.  For the tablets files it only compares the file path and not the value.  Currently when we read a tablet files we get a `Map<StoredTabletFile, DataFileValue>`.  When requireSame is called and told to check that the tablets files are the same it only checks the keys in the map are the same (it checks if the set of StoredTabletFile is the same).  Ideally it would check the entire map is the same in order to be more rigorous.

### PR DESCRIPTION
**Pull request by dlmarion**
Fixes #3522

### CODE REVIEW

**Comment by dlmarion** [comment]
Supercedes #4175

File: server/base/src/main/java/org/apache/accumulo/server/metadata/iterators/SetEqualityIterator.java [hunk 1]
```diff
@@ -45,16 +47,22 @@
 
 /**
  * This iterator exists to enable checking for set equality in a conditional mutation. It allows
- * comparing a set in a client process to a set encoded in column qualifiers within a tablet.
+ * comparing a set in a client process to a set encoded in column qualifiers within a tablet. If the
+ * "concat.value" options is supplied and is true, then the bytes from the Value will be added with
+ * a null byte separator.
```
Comment (DomGarguilo) [inline server/base/src/main/java/org/apache/accumulo/server/metadata/iterators/SetEqualityIterator.java:52]: Could do something like the following. This better allows for addition of more opts in the future and separates them from the description.
```suggestion
 * comparing a set in a client process to a set encoded in column qualifiers within a tablet.
 *
 * <h3>Options</h3>
 * <ul>
 * <li><b>concat.value:</b> If this option is supplied and is true, then the bytes from the Value
 * will be concatenated with a null byte separator.</li>
 * </ul>
```

File: server/base/src/test/java/org/apache/accumulo/server/SetEqualityIteratorTest.java [hunk 2]
```diff
@@ -150,8 +154,11 @@ public void testTabletWithOneFile() throws IOException {
     // Asserting the result
     assertEquals(new Key(tabletRow, family), setEqualityIteratorOneFile.getTopKey());
     // The iterator should produce a value that is equal to the expected value on the condition
-    var condition = SetEqualityIterator.createCondition(Collections.singleton(file1),
-        storedTabletFile -> storedTabletFile.getMetadata().getBytes(UTF_8), family);
+    var condition =
+        SetEqualityIterator.createCondition(tmOneFile.getFilesMap().entrySet(), entry -> {
+          return (entry.getKey().getMetadata() + SetEqualityIterator.VALUE_SEPARATOR
+              + entry.getValue().encodeAsString()).getBytes(UTF_8);
+        }, family, true);
```
Comment (DomGarguilo) [inline server/base/src/test/java/org/apache/accumulo/server/SetEqualityIteratorTest.java:161]: I was curious so I tried messing around with the newly added `concatValue` booleans and all three tests that use `createCondition` pass when  `concatValue` is true OR false. Not sure if that is expected or not. If so, is there a test case that we could add that tests that this field is enforced within `createCondition`?

**Review by cshannon — COMMENTED** [review]
Overall this looks pretty good but I'm wondering if we should do a slightly different approach. This change uses a boolean flag to indicate if the value has been concatenated so that the iterator knows to do the same and compare. That part is fine for the iterator, but the downside to this is that the encoder must properly encode the key, the value separate, and the value when creating the condition. This seems error prone as someone might incorrectly write the encoder.

I was thinking instead maybe a second createCondtion() helper method could just take two encoders as an argument instead of the flag (one for the key and one for the value) and then this implicitly implies to encode the value so the iterator could have it's flag set. The encoding logic could be re-used between the iteraor and condition creation. I created an example of this in a branch here: https://github.com/apache/accumulo/commit/170990409e0c8e6d2fb20e23b966cfd95af0b47b

For testing, I think it would be a good idea to add or update the SetEqualityIteratorTest to show that things are not equal if concatValue is set differently between two different conditions (which is a point that @DomGarguilo made). Basically just showing that both true and false work as intended.

Secondly, I think it would be good to update `AmpleConditionalWriterIT` as that is a good place to add another test. There's already tests for requiring the same files, it could be updated to verify it rejects mutations when the file values are different but the metadata is the same.

File: server/base/src/test/java/org/apache/accumulo/server/SetEqualityIteratorTest.java (hunk 2 above)
Comment (dlmarion) [reply server/base/src/test/java/org/apache/accumulo/server/SetEqualityIteratorTest.java:161]: ok, I will have to look at this and add more tests. Thanks!

**Comment by cshannon** [comment]
@dlmarion - A couple more things:

1. I am going to open a PR with my branch against your PR to make it easier to look at, but I noticed that the testFiles() test inside AmpleConditionalWriterIT is failing with this change here (without my changes as well) so I wanted to look at that.
2. I took a look at using BiFunction instead of two functions but I don't think that will work as that does the opposite of what we need. BiFunction takes 2 inputs and outputs 1 result but we need 1 input (the column) and to output 2 results (a key and value) that the iterator can concatenate transparently to the user when building the condition.

**Comment by cshannon** [comment]
I created a PR to your branch https://github.com/dlmarion/accumulo/pull/47

It fixes AmpleConditionalWriterIT as well, that test was failing because the changes here worked and the values were different. I updated the test to verify both the rejected and accepted case

File: server/base/src/main/java/org/apache/accumulo/server/metadata/iterators/SetEncodingIterator.java [hunk 3]
```diff
@@ -139,6 +161,7 @@ public boolean hasTop() {
   public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options,
       IteratorEnvironment env) throws IOException {
     this.source = source;
+    this.concat = Boolean.parseBoolean(options.getOrDefault(CONCAT_VALUE, Boolean.toString(false)));
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/metadata/iterators/SetEncodingIterator.java:164]: This option should always be set, so could throw an exception if its not present.  Following would throw an NPE, could do something else that creates an exception w/ a nice error message.

```suggestion
    this.concat = Boolean.parseBoolean(options.get(CONCAT_VALUE));
```

**Review by keith-turner — APPROVED** [review]

File: server/base/src/main/java/org/apache/accumulo/server/metadata/iterators/SetEncodingIterator.java (hunk 3 above)
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/metadata/iterators/SetEncodingIterator.java]: Added in a2413f5

**Pull request merged**